##part 1 metagenomic raw data processing
#1.1 microbial taxonomy and pathway abundance were generated by using MetaPhlAn2.0 and HUMAnN2.0, pipeline available via: https://github.com/GRONINGEN-MICROBIOME-CENTRE/Groningen-Microbiome/blob/master/Scripts/Metagenomics_pipeline_v1.md
#1.2 microbial antibiotic resistance and virulence gene abundance were generated by using ShortBRED, pipeline available via: https://huttenhower.sph.harvard.edu/shortbred

##part 2 statistics
#2.1 diversity, species, path, anti, vir abundance difference
library(microbiome)
sp=read.delim("../data/trackDC_110sample_238sp_metaphlan2.txt",sep = "\t",header = T,check.names = F)
sp=sp[1:104,]
sp=sp/rowSums(sp)
mean(rowSums(sp[,colSums(sp>0)>0.1*nrow(sp)]))
pheno=data.frame(t(abundances(t(sp),transform = "clr")),check.names = F)
sp=pheno[,colnames(sp[,colSums(sp>0)>0.1*nrow(sp)])]
pheno=read.delim("../trackDC_study1/table/data_pheno_pcoa_110samples.txt",header = T,sep = "\t")
pheno=pheno[row.names(sp),]
path=read.delim("../data/trackDC_110sample_398path_humann2.txt",sep = "\t",header = T,check.names = F)
tmp=data.frame(t(abundances(t(path),transform = "clr")),check.names = F)
path=tmp[,colSums(path>0)>0.1*nrow(path)]
path=path[row.names(pheno),]
diversity=read.delim("../data/trackDC_110sample_3diversity_index.txt",sep = "\t",header = T,check.names = F)
tmp=read.delim("../trackDC_study1/table/data_pheno_pcoa_110samples.txt",sep = "\t",header = T,check.names = F)[,1:2]
anti=read.delim("../data/trackDC_110sample_148antibiotic_gene_10percent.txt",sep = "\t",header = T,check.names = F)
anti=anti[row.names(pheno),]
virul=read.delim("../data/trackDC_110sample_55virul_gene_10percent.txt",sep = "\t",header = T,check.names = F)
virul=virul[row.names(pheno),]
data=pheno[,c(16,17,1:2,13,18:54)]
sp=sp[row.names(data),]
data=cbind(data,sp)
path=path[row.names(data),]
data=cbind(data,path)
anti=anti[row.names(data),]
data=cbind(data,anti)
virul=virul[row.names(data),]
data=cbind(data,virul)
result=data.frame(factor=NA,mean_control=NA,mean_rmt=NA,mean_rft=NA,P.kur=NA)
for(i in 3:ncol(data)){
  for(j in unique(data$day)){
    result=rbind(result,c(paste(colnames(data)[i],"_day",j,sep=""),mean(data[,i][which(data$day==j&data$group==1)],na.rm = T),mean(data[,i][which(data$day==j&data$group==2)],na.rm = T),mean(data[,i][which(data$day==j&data$group==3)],na.rm = T),try(kruskal.test(data[,i][which(data$day==j)]~data$group[which(data$day==j)])$p.value,silent = T)))
  }
  for(j in unique(data$group)){
    result=rbind(result,c(paste(colnames(data)[i],"_group",j,sep=""),mean(data[,i][which(data$day==15&data$group==j)],na.rm = T),mean(data[,i][which(data$day==35&data$group==j)],na.rm = T),mean(data[,i][which(data$day==56&data$group==j)],na.rm = T),try(kruskal.test(data[,i][which(data$group==j)]~data$day[which(data$group==j)])$p.value,silent = T)))
  }
}
result=result[-1,]
result=result[-grep("Error",result$P.kur),]
result=result[-grep("NaN",result$P.kur),]
result$Q.kur=p.adjust(result$P.kur,method = "BH")
write.table(result,file = "../trackDC_study1/table/result_group_differ_sp_path_anti_virul_pheno_fdr_kru.txt",quote = F,sep = "\t",row.names = F)
#2.2 micorbial changes associated with phenotypic changes
delta_calcu=function(tmp1,tmp2,vs){
  library(stringr)
  row.names(tmp1)=str_split_fixed(row.names(tmp1),"_",2)[,1]
  row.names(tmp2)=str_split_fixed(row.names(tmp2),"_",2)[,1]
  delta=tmp2[intersect(row.names(tmp1),row.names(tmp2)),]-tmp1[intersect(row.names(tmp1),row.names(tmp2)),]
  row.names(delta)=paste(row.names(delta),vs,sep = "_")
  return(delta)
} 
cor_spearman=function(microbiome,metabolites,name){
  microbiome=microbiome[intersect(row.names(microbiome),row.names(metabolites)),]
  metabolites=metabolites[row.names(microbiome),]
  pvals = matrix(NA,ncol = ncol(microbiome),nrow = ncol(metabolites))
  cors = matrix(NA,ncol = ncol(microbiome),nrow = ncol(metabolites))
  for(i in 1:ncol(microbiome)){
    for(j in 1:ncol(metabolites)){
      a=microbiome[is.na(microbiome[,i])==F&is.na(metabolites[,j])==F,i]
      b=metabolites[is.na(microbiome[,i])==F&is.na(metabolites[,j])==F,j]
      if(length(a)>2&length(b)>2){
        cor = cor.test(a,b,method = "spearman")
        pvals[j,i] = cor$p.value
        cors[j,i] = cor$estimate
      }
    }
  }
  qvals = matrix(p.adjust(pvals,method = "BH"),ncol = ncol(pvals))
  colnames(qvals) = colnames(microbiome)
  rownames(qvals) = colnames(metabolites)
  colnames(cors) = colnames(microbiome)
  rownames(cors) = colnames(metabolites)
  ind = which(pvals <=1,arr.ind = T)
  association = data.frame(metabolites = colnames(metabolites)[ind[,1]],microbiome = colnames(microbiome)[ind[,2]],Cor = cors[ind],Pval = pvals[ind],Qval = qvals[ind],stringsAsFactors = F)
  association$name=paste(association$metabolites,association$microbiome,sep = "_with_")
  colnames(association)[1:2]=name
  qvals[is.na(qvals)]=1
  cors=cors[row.names(qvals),colnames(qvals)]
  pvals[pvals<0.05]="."
  pvals[qvals<0.05]="*"
  pvals[pvals>0.05]=NA
  qvals[qvals<0.05]="*"
  qvals[qvals>0.05]=NA
  return(list(association,cors,qvals,pvals))
}
pheno_15_35=delta_calcu(pheno[grep("_15",row.names(pheno)),],pheno[grep("_35",row.names(pheno)),],"15_35")
pheno_15_56=delta_calcu(pheno[grep("_15",row.names(pheno)),],pheno[grep("_56",row.names(pheno)),],"15_56")
microbial_15_35=delta_calcu(microbial[grep("_15",row.names(microbial)),],microbial[grep("_35",row.names(microbial)),],"15_35")
microbial_15_56=delta_calcu(microbial[grep("_15",row.names(microbial)),],microbial[grep("_56",row.names(microbial)),],"15_56")
ass_g1_35=cor_spearman(pheno_15_35[grep("g1",row.names(pheno_15_35)),],microbial_15_35[grep("g1",row.names(microbial_15_35)),],c("microbe","pheno"))[[1]]
colnames(ass_g1_35)[3:5]=paste(colnames(ass_g1_35)[3:5],"g1",sep = "_")
ass_g2_35=cor_spearman(pheno_15_35[grep("g2",row.names(pheno_15_35)),],microbial_15_35[grep("g2",row.names(microbial_15_35)),],c("microbe","pheno"))[[1]]
colnames(ass_g2_35)[3:5]=paste(colnames(ass_g2_35)[3:5],"g2",sep = "_")
ass_g3_35=cor_spearman(pheno_15_35[grep("g3",row.names(pheno_15_35)),],microbial_15_35[grep("g3",row.names(microbial_15_35)),],c("microbe","pheno"))[[1]]
colnames(ass_g3_35)[3:5]=paste(colnames(ass_g3_35)[3:5],"g3",sep = "_")
ass=merge(ass_g1_35[,3:6],ass_g2_35[,3:6],by="name",all = F)
ass=merge(ass,ass_g3_35,by="name",all = F)
ass=ass[,c(8:9,2:7,10:12)]
ass=ass[which(ass$Pval_g1<0.01|ass$Pval_g2<0.01|ass$Pval_g3<0.01),]
write.table(ass,file = "../trackDC_study1/table/result_ass_delta_pheno_microbe_15_35.txt",quote = F,sep = "\t",row.names = F)
ass_g1_35=cor_spearman(pheno_15_56[grep("g1",row.names(pheno_15_56)),],microbial_15_56[grep("g1",row.names(microbial_15_56)),],c("microbe","pheno"))[[1]]
colnames(ass_g1_35)[3:5]=paste(colnames(ass_g1_35)[3:5],"g1",sep = "_")
ass_g2_35=cor_spearman(pheno_15_56[grep("g2",row.names(pheno_15_56)),],microbial_15_56[grep("g2",row.names(microbial_15_56)),],c("microbe","pheno"))[[1]]
colnames(ass_g2_35)[3:5]=paste(colnames(ass_g2_35)[3:5],"g2",sep = "_")
ass_g3_35=cor_spearman(pheno_15_56[grep("g3",row.names(pheno_15_56)),],microbial_15_56[grep("g3",row.names(microbial_15_56)),],c("microbe","pheno"))[[1]]
colnames(ass_g3_35)[3:5]=paste(colnames(ass_g3_35)[3:5],"g3",sep = "_")
ass=merge(ass_g1_35[,3:6],ass_g2_35[,3:6],by="name",all = F)
ass=merge(ass,ass_g3_35,by="name",all = F)
ass=ass[,c(8:9,2:7,10:12)]
ass=ass[which(ass$Pval_g1<0.01|ass$Pval_g2<0.01|ass$Pval_g3<0.01),]
write.table(ass,file = "../trackDC_study1/table/result_ass_delta_pheno_microbe_15_56.txt",quote = F,sep = "\t",row.names = F)
#2.3 microbial co-abundance networks
#sparcc (Python) for network inference
python SparCC.py $INPUT_PATH --cor_file=$OUTPUT_PATH/real_cors_sparcc_sp.txt
python MakeBootstraps.py $INPUT_PATH -n $BOOT_ITER -t permutation_#.txt -p $OUTPUT_PATH/Resamplings/
for i in 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99
do
python SparCC.py $OUTPUT_PATH/Resamplings/permutation_$i.txt --cor_file=$OUTPUT_PATH/Bootstraps/sim_cor_$i.txt
done
python PseudoPvals.py $OUTPUT_PATH/real_cors_sparcc_sp_fg.txt $OUTPUT_PATH/Bootstraps/sim_cor_#.txt $BOOT_ITER -o $OUTPUT_PATH/pvals_two_sided_sp_fg.txt -t two_sided
library(meta)
TE=c(cor_group1,cor_group2,cor_group3) #coefficient of certain edge
TEse=c(sqrt((1-cor_group1^2)/(12-2)),sqrt((1-cor_group2^2)/(12-2)),sqrt((1-cor_group3^2)/(12-2)))  #calculate se of coefficient
p_heterogeneity=metagen(TE,TEse)$pval.Q
TE=c(0.5*log((1+cor_group1)/(1-cor_group1)),0.5*log((1+cor_group2)/(1-cor_group2)),0.5*log((1+cor_group3)/(1-cor_group3))) #transfer spearman coefficient to z-score (based on Park E. & Lee Y. J., Communications in Statistics-Simulation and Computation, 2001.)
TEse=c(1/sqrt(group1_n-3),1/sqrt(group2_n-3),1/sqrt(group3_n-3)) #calculate se of z-score, group1_n represent non-zero pairs of certatin in group1 (based on Park E. & Lee Y. J., Communications in Statistics-Simulation and Computation, 2001.)
p_heterogeneity=metagen(TE,TEse)$pval.Q
